src/sagas/nanoContract.js

```diff
@@ -245,28 +247,21 @@ export async function fetchHistory(req) {
     throw new Error('Failed to fetch nano contract history');
   }
 
-  /* TODO: Make it run concurrently while guaranting the order.
-  /* see https://github.com/HathorNetwork/hathor-wallet-mobile/issues/514
-   */
-  const historyNewestToOldest = new Array(rawHistory.length)
-  for (let idx = 0; idx < rawHistory.length; idx += 1) {
-    const rawTx = rawHistory[idx];
-    const network = wallet.getNetworkObject();
+  const network = wallet.getNetworkObject();
+  // As isAddressMine call is async we collect the tasks to avoid
+  // suspend the iteration with an await.
+  const isMineTasks = [];
+  // Translate rawNcTxHistory to NcTxHistory and collect isAddressMine tasks
+  const historyNewestToOldest = rawHistory.map((rawTx) => {
```

Comment:
Since every iteration of this loop will create a "task" to be awaited later, why don't you just return promises in this map and await them? something like:

```
const historyTransformPromises = rawHistory.map(async (rawTx) => {
  const caller = addressUtils.getAddressFromPubkey(rawTx.nc_pubkey, network).base58;
  const actions = rawTx.nc_context.actions.map((each) => ({
    type: each.type, // 'deposit' or 'withdrawal'
    uid: each.token_uid,
    amount: each.amount,
  }));

  const isMine = await utils.isAddressMine(wallet, caller, useWalletService);

  const tx = {
    txId: rawTx.hash,
    timestamp: rawTx.timestamp,
    tokens: rawTx.tokens,
    isVoided: rawTx.is_voided,
    ncId: rawTx.nc_id,
    ncMethod: rawTx.nc_method,
    blueprintId: rawTx.nc_blueprint_id,
    firstBlock: rawTx.first_block,
    caller,
    actions,
    isMine,
  };
  return tx;
});

const historyNewestToOldest = await Promise.all(historyTransformPromises);
```

This would also avoid iterating the loop twice

Reply:
I believe we should start to measure performance to really answer this kind of question. But in the lack of data there is a catch in your proposal, the `await` provokes the block code to be suspended and add other operations as a cost. In my proposal the iteration runs from start to end without any suspension in both iterations despite to be two of them, there is only one suspension in the script `Promise.all` and of course the suspensions that happens in the `utils.isAddressMine` calls, but they happen out of the iteration context, which probably make it quicker. But his mental mode needs a performance test to be proof. Maybe we didn't deliver any benefit at all, but doesn't it makes sense?

Reply:
I don't believe there will be any noticeable difference in performance, and even if there is, it will most likely be irrelevant.

Regarding the 'suspensions' you mentioned, I'm not sure I fully understand what you mean. JavaScript execution isn't blocked at all when awaiting promises. When the utils.isAddressMine promise is awaited, the promise returned by the lambda passed to map is paused until it resolves, but execution continues elsewhere in the code, including other iterations of the map. Once the promise resolves, everything after the await is placed back into the event loop and executed.

The JavaScript event loop itself isn't suspended by awaiting promises. The JavaScript engine efficiently manages asynchronous operations.

Given that, I think the main focus here should be on code clarity. In my view, the version I suggested is clearer, as it handles everything in one go without the need for additional mutations in a separate loop

Reply:
>I don't believe there will be any noticeable difference in performance, and even if there is, it will most likely be irrelevant.

I've tested it and in fact there aren't a significant difference, despite mine proposition be slightly better, but almost not noticeable.

```log
❯ node benchmark.js
Function 1 { avg: 602.44054917, ops: 1.6599148270775086 }
Function 2 { avg: 602.5024887400015, ops: 1.6597441814576985 }
```

I will not post here the benchmark ugly code, but if you want I share with you.

I'm adopting your proposition because of the clarity gain.